### PR TITLE
Fix android toolchain & arm platform matching.

### PR DESCRIPTION
**Pull request by yeswalrus**
Removes a bunch of cruft from our android toolchain files and allows use of cmake's standard platform support. Also fixes a detection error with others who may already be using a saner toolchain file.
Also included are fixes for issues reparsing toolchain files (replacing standard_project with an override) and some other cleanup.

To verify, see https://github.com/leapmotion/autowiring/tree/fix-android-toolchain

- [x] verify autowiring on windows
- [x] on mac
- [x] on android64
- [x] Test with Unity native plugin (android32)

### CODE REVIEW

**Review by jdonald — COMMENTED** [review]
Tests pass for me so far in Android 64-bit, but with Android 32-bit I get this:

```
cmake -DCMAKE_ANDROID_STANDALONE_TOOLCHAIN=/opt/android-standalone-toolchain -DCMAKE_TOOLCHAIN_FILE=../standard/toolchain-android.cmake -G Ninja  ..
CMake Error at standard/toolchain-android.cmake:19 (message):
  ANDROID_NDK_ROOT could not be located.  Have you installed the NDK?
Call Stack (most recent call first):
  /usr/local/share/cmake-3.8/Modules/CMakeDetermineSystem.cmake:85 (include)
  CMakeLists.txt


CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
```

Odd considering the two paths now aren't that different.

**Comment by jdonald** [comment]
I was on the wrong branch (override-project) with the above test, but when I move to this one and toolchain-android32.cmake I get a different problem in Android 32-bit: It defaults to the x86_64 toolchain.

**Comment by yeswalrus** [comment]
@jdonald Fixed, was a silly error

File: standard/toolchain-android32.cmake [hunk 1]
```diff
@@ -1,2 +1,3 @@
 set(CMAKE_ANDROID_ARCH_ABI armeabi-v7a)
 set(CMAKE_ANDROID_ARM_MODE ON)
+include(toolchain-android.cmake)
```
Comment (gittyupagain) [inline standard/toolchain-android32.cmake]: It appears that this is missing a trailing newline character.

**Comment by yeswalrus** [comment]
added Fixup commits for missing newline & absolute path to the android toolchain file.

**Comment by jdonald** [comment]
This works (at least builds and runs AutowiringTest) in 32-bit for me now, but I see it also adds `-march=armv7-a` i.e. the bug that broke our Android Unity API support for two months.

Some possibilities:
* I see there's additional auto-generated compiler options now including `-mfpu=vfpv3-d16 -funwind-tables -no-canonical-prefixes`. It would be great if it was only when we lacked these options that problems arose if adding `-march=armv7-a`
* Perhaps it's okay to use this in autowiring, leapserial, etc. and the flag only causes problems when directly used downstream.
* It could be the *mixing* of libraries with and without `-march=armv7-a` that causes the hang. If that's the cause, we should be wary of adding the flag here as that will require updating many other 3rd-party libraries, not just the ones using leapmotion/standard

With a little testing with can figure out the best course of action here.

**Comment by yeswalrus** [comment]
This might be a bit tricky then. I'm noticing that our old toolchain files were not setting either -mthumb or -march at all, and cmake's built in android support will *always* set at least some of those flags. It's hard to determine exactly what our old flags were equivalent to, but setting -march=armv5te and -mthumb is cmake's default behavior with our version of the ndk.  would that be acceptable?

**Comment by hanzou-x** [comment]
It just depends on which configuration leads to or avoids the Unity hang. It should be clear from testing, and even if we put a hack in a `REMOVE` to remove the `-march=armv7-a` we should test this regardless (as it changes other flags).

**Comment by jdonald** [comment]
It just depends on which configuration leads to or avoids the Unity hang. It should be clear from testing, and even if we put a `REMOVE` hack in to remove the `-march=armv7-a` we should test this regardless (as it changes other flags).

**Comment by yeswalrus** [comment]
ok, from doing some runs with gcc -v, it appears that the old behavior was equivalent to -march=armv5te. As such, I'm setting it to use that behavior by default if no alternative is specified, but for the sake of generality I'm still allowing it to be overridden with -D on the command line.

**Comment by yeswalrus** [comment]
For compatibility's sake, we'll set the ABI to armeabi if it's not set, which will be the equivalent to defaulting to our old behavior.

**Comment by yeswalrus** [comment]
It is worth noting that in experiments I ran, linking with a version of autowiring compiled with -march=armv7-a does *not* cause any problems with hangs in Unity.

**Comment by yeswalrus** [comment]
@jdonald I've tested this locally as working by compiling autowiring & LeapC & running a unity scene

**Comment by jdonald** [comment]
Great! So mixing armv7-a code with non-armv7-a code alone does not directly cause the hang.

Although the proposal here is not only to change the `march` for autowiring, but also for leapserial+leapipc+leapresource+leaphttp. I suspect it'll work fine (the hang likely directly caused in downstream LeapC code), but should be quick to verify now that your flow is set up?

**Comment by yeswalrus** [comment]
Verified that this branch does not cause the android32 hang (though the service I was running didn't show hands in any apps)

**Comment by jdonald** [comment]
Was that with the change applied to both autowiring and LeapIPC? Wondering if LeapIPC issues could be related to not seeing hands, unless that happens your baseline too  due to some unrelated issue with the setup.

**Comment by yeswalrus** [comment]
not seeing hands was baseline

**Comment by yeswalrus** [comment]
Confirmed that linking with compiled versions of the libraries *does* cause hands to fail to appear. This appears to be conclusive proof that armeabi and armeabi-v7a are not compatible for our purposes. I'll stop treating them as the same in the detection code and set the default behavior to match our old stuff.

**Comment by yeswalrus** [comment]
Verified that this works correctly with android32 leapc

**Comment by yeswalrus** [comment]
Confirmed that building LeapC against libraries built with this PR works correctly on android32/Unity.

**Comment by jdonald** [comment]
Great!

**Comment by jdonald** [comment]
Note I was testing these with CMake 3.10.2. It's possible that any repos that merge in the latest standard will now require this and should have their `.travis.yml` steps updated accordingly.

File: standard/toolchain-android64.cmake [hunk 2]
```diff
@@ -1,4 +1,3 @@
-set(ANDROID_NDK_TOOL_PREFIX aarch64-linux-android)
-set(_ndk_suffix -4.9-aarch64)
-set(CMAKE_SYSTEM_PROCESSOR aarch64)
+set(CMAKE_ANDROID_ARCH_ABI arm64-v8a)
 include(${CMAKE_CURRENT_LIST_DIR}/toolchain-android.cmake)
+
```
Comment (jdonald) [inline standard/toolchain-android64.cmake:3]: Should have spotted this earlier, but now that standard is getting merged into other repos I'm noticing this stray extra newline.

**Pull request merged**